### ISSUE TITLE
Equivilant of RemoteAddr on all Channels.

### ISSUE THREAD

**Issue by DavidStrickland0** [microsoft/botbuilder-dotnet#3656]
**Is your feature request related to a problem? Please describe.**
I need to be able to tell where an Activity is from. There are any number of reasons, almost anything an existing website does with RemoteAddr would apply. Throttling, Tailoring, Tracking etc. but one example is we'd like to be able to reverse geolocate the user to make sure to use the QnA thats specific for their country. If the location cant be identified then a generic QnA is used.

**Describe the solution you'd like**
Dependent on which Channel is being used the Azure "Web App Bot" must have specific information about the requester. In the WebChannel for example Azure "Web App Bot" has the RemoteAddr Header. Presumably other channels have similar info. This info should be passed along in the ChannelData property of the Activity to the Bot App Service. 

**Describe alternatives you've considered**
At first glance ServiceUrl appears to provide this information but once the bot is deployed to Azure ServiceUrl always returns as  “https://directline.botframework.com/” which is presumably the direct upstream service IE the Azure "Web App Bot".

**Additional context**
There is a Stackoverflow from back in 2017 that describes a work around. Unfortunately this only works for the WebChannel. 
https://stackoverflow.com/questions/42825048/bot-framework-get-the-serviceurl-of-embedded-chat-control-page

[enhancement]

**Comment by cleemullins** [comment]
@DavidStrickland0 We don't have any currently planned work in this area. If you describe your scenario in more detail, me may be able to provide suggestions for a specific workaround.

**Comment by DavidStrickland0** [comment]
Not sure how else to describe it. If someones attaching to my Bot via any channel that allows anonymous connection I need to have some way to identify them. GeoLocate them, Track them, Store Data about them, make educated guesses about who they are and what they need. On a website you do that using their IP Address and or the Referrer and whatever other headers you have on the HttpRequest. If the user talked directly to my Bot I'd have all that by looking at those http headers. With the Microsoft Azure Web App Bot in between me and my user performing pass through I can't get any header information. Relying on the bot consumer to write their WebClient in such a way as to send me that info in Events isnt reliable. Most of the channels are authenticated channels but any channel that allows anonymous connections your going to want \ need as much info as you can get about the incoming connection. You don't have a user context so you have to fall back on the data that you do have …. which at the moment is none at all.

**Comment by DavidStrickland0** [comment]
Perhaps our biggest concern atm is Scraping. On our websites its a constant war between the data scrapers and the site administrators. If we create a bot that allows via dialog access to the same data we have on the site its a new frontline in the war. Currently we are constantly analysing traffic. Blocking \ Throttling or misinforming requests we Identify as performing questionable activity while at the same time attempting to keep channels open for legitimate users. We need as much data about the incoming request as we can get to segregate legitimate bot users from automated data scrapers. 
On the other side of the equation we want our Bot to be as intelligent as possible. We want to feed as much data as we can into our AI Models so we can to get an intuitive bot that can predict what the user wants. The more data we have about the incoming request the better we can predict what the user needs. In this case browser version\ OS Version  would even be of value any and ever data feature we can feed into the model may well prove to be the key data point. Knowing that the  Microsoft Azure Web App Bot is swallowing that data and not passing it through to the Bot is very frustrating.

**Comment by cleemullins** [comment]
@garypretty, any update on this?

**Comment by garypretty** [comment]
@DavidStrickland0 Are you able to provide me with the different anonymous channels that you are currently / planning to use?  It's also not clear to me if you have control of the web chat implementations for your bot - can you clarify?

**Comment by DavidStrickland0** [comment]
@garypretty At this point the only anonymous channel we are offering is the Web Channel that however will change as our clients request additional avenues of engagement. If there is another Anonymous channel that Microsoft supports out of the box I'm fairly certain we'll end up having to support it eventually.

As for control of the Web Chat Implementations given its an open platform that anyone can write a WebChat Implementation for it's not something I can say yes to. Granted we have a web chat implementation that our UI team has created which we plan to offer. However as a backend developer I never trust the source and have to assume that every scraper that wants to automate the bot and pull data and eventually every consumer is going to use their own custom implementation. They may start using what we offer but it wont stay that way for long when they realize they can customize it and write their own. We've got security by obscurity at the moment. I doubt it will last.

**Comment by garypretty** [comment]
@DavidStrickland0 just so I am clear, you have created the bot, but then are allowing other developers to implement their own web chat interface to talk to the bot?  If you are building your own web chat interface then you should be able to secure this so that only your web chat implementation can communicate with the bot.  Are you using the token API approach?  You can find the links to doc for that and the web chat authentication at https://docs.microsoft.com/en-us/azure/azure-monitor/app/overview-dashboard#troubleshooting

**Comment by DavidStrickland0** [comment]
At the end of the day every web channel bot is an API sitting behind an Api at
https://directline.botframework.com/v3/directline/conversations. 

Not certain how a Token Api Approach will help. We are using it but that really just gives us tracking. If an Anonymous user can get a token all it does is add one more step in the process of gaining access... Get the token then access the bot. We can put some control on how many tokens we issue etc. But in those cases all I have is the Token Request not what is happening in the bot using that token...... or is there a way to see the Token from withen the bot code so I can aggregate that data back to the Token Request? 

IMHO: Every Bot Developer should be assuming that any Client can access their bot. Once the Microsoft Bot Framework goes live I'm sure we'll start seeing Bot Compilation Sites and Universal Bot Apps that aggregate Multiple bots into a single interface. You just add a Bot ID to your Contact List and by providing the Bot Id it gets added to a Universal bot phone book thats crowd sourced. And of course 101 other Unexpected applications no ones even thinking of at the moment.

**Comment by garypretty** [comment]
Having confirmed with other members of the team internally, right now you will only receive the information received via the HTTP request, which is going to vary between channel / networking. You can, as you mentioned, use the extensibility options of the latest web chat control to send additional information along on the incoming activity. You could also use something like the HTML5 capabilities within the browser to get the user's location and send that too, but this would still depend on the user agreeing to provide their location when asked.  This is going to be the same situation with other channels too - for example, if you are using Facebook, then you can explicitly request location using built in Facebook capabilities. Again though, this relies on the user agreeing to provide it.

For most of the channels, other than web chat, you will get some sort of UUID from the channel (even if it doesn't identify the user explicitly). Ultimately, if users are using your bot anonymously, then the only way to be sure that they are a 'valid' user would be to authenticate them in some way, which obviously is not feasible / possible in every case.

I would definitely recommend you look into using the web chat extensibility, as I said. You can find a sample of piggybacking data on the back of outgoing activities at https://github.com/microsoft/BotFramework-WebChat/tree/master/samples/04.api/b.piggyback-on-outgoing-activities

We really appreciate your feature request in this area, but right now we have no plans to implement anything additional within this area.  I am going to close out this issue now, but if you are struggling to implement any of the additional features for your web chat implementation, then please raise an issue in the webchat repo (https://github.com/microsoft/BotFramework-WebChat).

**Comment by DavidStrickland0** [comment]
Gary thanks for looking at this. 
After looking again at the docs for webchat Authentication https://docs.microsoft.com/en-us/azure/bot-service/rest-api/bot-framework-rest-direct-line-3-0-authentication?view=azure-bot-service-4.0#security-considerations

I realized I was missing the fact that the Token Request returns both a token and a ConversationID. This presumably will allow me to tie the conversation back to the token request which will let me have access to any data that posted with the TokenRequest which would include then the RemoteAddr that the token request came from. Read right past this several times before I realized what it potentially means. Figured I'd post it up in case someone else stumbles on this later.

Thanks Again
Dave
